owners/src/reviewer_selection.js

```diff
@@ -0,0 +1,186 @@
+    return Object.entries(fileTreeMap)
+      .filter(([filename, tree]) => tree.hasOwner(reviewer))
+      .map(([filename, tree]) => filename);
+  }
+
+  /**
+   * Identifies the reviewer(s) who can approve the most files.
+   *
+   * @private
+   * @param {!Object<string, string[]>} reviewerFilesMap map from reviewer
+   *     usernames to the list of files they own.
+   * @return {string[]} list of reviewer usernames.
+   */
+  static _reviewersWithMostFiles(reviewerFilesMap) {
+    const mostFilesOwned = Math.max(
+      ...Object.values(reviewerFilesMap).map(files => files.length)
+    );
+    return Object.entries(reviewerFilesMap).filter(
+      ([reviewer, files]) => files.length == mostFilesOwned
+    );
+  }
+
+  /**
+   * Picks the best reviewer-fileset pair to satisfy the deepest rules.
+   *
+   * @private
+   * Chooses randomly from the set of best possible reviewers to allow even
+   * distribution of reviews.
+   *
+   * @param {!FileTreeMap} fileTreeMap map from filenames to ownership subtrees.
+   * @return {ReviewerFiles} tuple of a reviewer username and the files they
+   *     own.
+   */
+  static _pickBestReview(fileTreeMap) {
+    const reviewerSet = this._findPotentialReviewers(fileTreeMap);
+    const reviewerFilesMap = {};
+    reviewerSet.forEach(reviewer => {
+      reviewerFilesMap[reviewer] = this._filesOwnedByReviewer(
+        fileTreeMap,
+        reviewer
+      );
+    });
+    const bestReviewers = this._reviewersWithMostFiles(reviewerFilesMap);
+    return bestReviewers[Math.floor(Math.random() * bestReviewers.length)];
+  }
+
+  /** Part 3 **/
+
+  /**
+   * Builds the map from filenames to ownership subtrees.
+   *
+   * @param {string[]} filenames list of changed files.
+   * @param {!OwnersTree} ownersTree root node of the ownership tree.
+   * @return {FileTreeMap} map from filenames to nearest ownership subtrees.
+   */
+  static buildFileTreeMap(filenames, ownersTree) {
+    const fileTreeMap = {};
+    filenames.forEach(filename => {
+      fileTreeMap[filename] = ownersTree.atPath(filename);
+    });
+    return fileTreeMap;
+  }
+
+  /**
+   * Picks a set of reviewers to approve the PR.
+   *
+   * @throws {Error} if the algorithm fails to select reviewers.
+   * @param {!FileTreeMap} fileTreeMap map from filenames to ownership subtrees.
+   * @return {ReviewerFiles[]} list of reviewers and the files they cover, in
+   *     decreasing order of ownership depth.
+   */
+  static pickReviewers(fileTreeMap) {
+    const reviews = [];
+
+    while (Object.entries(fileTreeMap).length) {
+      const bestReview = this._pickBestReview(fileTreeMap);
+      if (!bestReview) {
+        // This should be impossible unless there is no top-level OWNERS file.
+        throw Error('Could not select reviewers!');
+      }
+
+      const [bestReviewer, coveredFiles] = bestReview;
+      reviews.push([bestReviewer, coveredFiles]);
+      coveredFiles.forEach(filename => delete fileTreeMap[filename]);
```

Comment:
This feels… off? I dunno why, this might be perfectly normal. Just never seen this done before this way. Maybe somebody who feels much more comfortable with JS than me can chime in

Reply:
What feels off? Removing entries from a dict?

Reply:
Wrapped the lambda in braces